### PR TITLE
feat(new source): ecs/fargate metrics source

### PR DESCRIPTION
**Pull request by drunkirishcoder**
implements #4683. still a work in progress. need to find a reference implementation to finalize which metric is a gauge and which is a counter. can't find any good documentation on the docker container stats format. in the mean time, would love to get some feedback on if this general approach is sound.

### CODE REVIEW

File: src/internal_events/aws_ecs_metrics.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+impl InternalEvent for AwsEcsMetricsReceived {
+    fn emit_logs(&self) {
+        debug!(message = "Scraped events.", ?self.count);
+    }
+
+    fn emit_metrics(&self) {
+        counter!(
+            "events_processed", self.count as u64,
```
Comment (jamtur01) [inline src/internal_events/aws_ecs_metrics.rs:19]: We've updated all our metrics to basically use the Prometheus style metric naming (see https://github.com/timberio/vector/pull/4647).

File: src/internal_events/aws_ecs_metrics.rs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            "component_kind" => "source",
+            "component_type" => "aws_ecs_metrics",
+        );
+        counter!(
+            "bytes_processed", self.byte_size as u64,
+            "component_kind" => "source",
```
Comment (jamtur01) [inline src/internal_events/aws_ecs_metrics.rs:25]: You shouldn't need component_type and _kind anymore.

**Comment by jamtur01** [comment]
For the Docker container stats format - the API doesn't have the metric types defined (https://docs.docker.com/engine/api/v1.40/#operation/ContainerStats) so I tend to go off what caAdvisor outputs and correlating that with the API (https://github.com/google/cadvisor/blob/master/docs/storage/prometheus.md).

**Comment by drunkirishcoder** [comment]
Yeah it's the same docker container stats.

**Comment by drunkirishcoder** [comment]
Sorry been a little busy, should be able to get back on this in the coming week!

**Comment by jamtur01** [comment]
@drunkirishcoder Thanks!

File: src/internal_events/aws_ecs_metrics.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+impl InternalEvent for AwsEcsMetricsRequestCompleted {
+    fn emit_logs(&self) {
+        debug!(message = "Request completed.");
+    }
+
+    fn emit_metrics(&self) {
+        counter!("requests_completed_total", 1);
+        histogram!("request_duration_nanoseconds", self.end - self.start,
```
Comment (jszwedko) [inline src/internal_events/aws_ecs_metrics.rs:36]: I believe this should be `request_duration_nanoseconds` to match our naming conventions (based on https://prometheus.io/docs/practices/naming/).

File: src/sources/aws_ecs_metrics/mod.rs [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+}
+
+fn aws_ecs_metrics(
+    url: String,
+    interval: u64,
+    shutdown: ShutdownSignal,
+    out: Pipeline,
+) -> super::Source {
+    let out = out
+        .sink_map_err(|e| error!("error sending metric: {:?}", e))
```
Comment (jszwedko) [inline src/sources/aws_ecs_metrics/mod.rs:94]: We've started to update these per https://github.com/timberio/vector/blob/master/CONTRIBUTING.md#logging-style

File: src/sources/aws_ecs_metrics/mod.rs [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+        .sink_compat();
+    let task = tokio::time::interval(Duration::from_secs(interval))
```
Comment (jszwedko) [inline src/sources/aws_ecs_metrics/mod.rs:96]: This is fine, but I actually find the way the scraping is setup in `host_metrics` a bit easier to read:

https://github.com/timberio/vector/blob/45cb856d7645ec0918b4ab051d5138bd5842e8c1/src/sources/host_metrics.rs#L155-L178

File: src/sources/aws_ecs_metrics/mod.rs [hunk 6]
```diff
@@ -0,0 +1,584 @@
...
+        .take_until(shutdown)
+        .map(move |_| {
+            let client = Client::new();
+
+            let request = Request::get(&url)
+                .body(Body::empty())
+                .expect("error creating request");
+
+            let start = Instant::now();
+            let url2 = url.clone();
```
Comment (jszwedko) [inline src/sources/aws_ecs_metrics/mod.rs:106]: I'm taking this was necessary, but I'm not seeing why.

File: src/sources/aws_ecs_metrics/mod.rs [hunk 7]
```diff
@@ -0,0 +1,584 @@
...
+                .flatten()
+        })
+        .flatten()
+        .forward(out)
+        .inspect(|_| info!("finished sending"));
+
+    Box::new(task.boxed().compat())
+}
+
+#[cfg(feature = "sinks-prometheus")]
```
Comment (jszwedko) [inline src/sources/aws_ecs_metrics/mod.rs:169]: Given the parser tests, I think making these ones less specific are OK. It'd be nice to not depend on `sinks-prometheus` so the module can be built and tested independently.

You can see an example in the `apache_metrics` tests: https://github.com/timberio/vector/blob/45cb856d7645ec0918b4ab051d5138bd5842e8c1/src/sources/apache_metrics/mod.rs#L262-L480 which just test higher level conditions and rely on the parser tests for more detailed assertions.

**Review by jszwedko — COMMENTED** [review]
This is looking good! Thanks @drunkirishcoder 

I left some comments below, but nothing major.

Regarding choosing the metadata endpoint version, I think it'd be nice to try to automatically use the newest version available. That is, make `version` in the config optional and:

* If `$ECS_CONTAINER_METADATA_URI_V4` is set, use that
* If `$ECS_CONTAINER_METADATA_URI` is set, use that
* Fall back to `169.254.170.2/v2/stats`

We can still let users explicitly set a `version` field if they want which would have the behavior you currently have.

Additionally, a common pattern we have is to allow setting the `endpoint` directly in the config. Setting this would require the `version` field as well so we know which paths to suffix. This can be useful when testing the component (eliminating the need to modify the environment as you have now) or if the user is accessing the metadata in a non-standard way (like through a proxy).

**Comment by drunkirishcoder** [comment]
I like your suggestion. I think it's best to eliminate the version completely and go through the auto discovery logic by default when no explicit endpoint is specified. I will go through rest of the comments tomorrow.

File: src/internal_events/aws_ecs_metrics.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (drunkirishcoder) [reply src/internal_events/aws_ecs_metrics.rs:36]: I don't think this is going to work. From the doc

> By default, histogram! will accept a u64 directly or a Duration, which uses the nanoseconds total as the converted value.

If I do the conversion from nanosecond to second then convert to u64, you will lose precision on the measurement.

[the remark this answers is not in the document]
Comment (jszwedko) [reply src/internal_events/aws_ecs_metrics.rs:36]: :+1: you are right, I see we are using `nanoseconds_total` in other places as well. I was thinking the histogram could take an `f64`.

We should aim to report all duration values as nanoseconds then. The most important bit is just consistency to make it easy to compare metrics.

[the remark this answers is not in the document]
Comment (drunkirishcoder) [reply src/internal_events/aws_ecs_metrics.rs:36]: Shouldn't be `total` though right? Cuz it's a histogram. Here's the same metric for prometheus source.

https://github.com/timberio/vector/blob/45cb856d7645ec0918b4ab051d5138bd5842e8c1/src/internal_events/prometheus.rs#L37

File: src/sources/aws_ecs_metrics/mod.rs (hunk 5 above)
Comment (drunkirishcoder) [reply src/sources/aws_ecs_metrics/mod.rs:96]: OK, I will give it a try after I finish addressing the other comments. I agree the new impl is more readable.

File: src/internal_events/aws_ecs_metrics.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jszwedko) [reply src/internal_events/aws_ecs_metrics.rs:36]: You're right :smile: just `request_duration_nanoseconds`

**Comment by drunkirishcoder** [comment]
@jamtur01 found a source that has the docker container stats documented. https://docs.signalfx.com/en/latest/integrations/agent/monitors/docker-container-stats.html.

**Comment by jamtur01** [comment]
@drunkirishcoder Nice! That'll be useful.

File: src/sources/aws_ecs_metrics/mod.rs (hunk 6 above)
Comment (drunkirishcoder) [reply src/sources/aws_ecs_metrics/mod.rs:106]: make it a ref and avoided cloning.

**Comment by drunkirishcoder** [comment]
The work on this is complete. I cross-checked against the signalfx [docs](https://docs.signalfx.com/en/latest/integrations/agent/monitors/docker-container-stats.html) to make sure I have the right metric types. All comments are addressed as well. The only thing I can think of is I did not rename all the metrics to follow the prometheus naming convention. Not sure what the thoughts on that is. Would that be a worthwhile effort?

**Comment by jamtur01** [comment]
@drunkirishcoder That'd be great if you could. We've settled on that as the standard.

**Comment by drunkirishcoder** [comment]
OK I may need to do the parsing a little differently then. Also some names may end up kinda weird when translated, like `memory.stats.total_pgfault` will probably be named `memory_total_page_faults_total` or something like that.

**Comment by drunkirishcoder** [comment]
OK tried my best to make the names nice and clean. Have a look, thanks. Also failed checks seems to be unrelated.

**Review by jamtur01 — APPROVED** [review]
LGTM

**Review by jszwedko — COMMENTED** [review]
This looks great overall, thanks @drunkirishcoder !

I left a few last comments, but also, this appears to give me a parsing error when I went to test it out with: https://github.com/awslabs/amazon-ecs-local-container-endpoints

```
$ docker run -d -v /var/run:/var/run -p 9080:80 --network aws amazon/amazon-ecs-local-container-endpoints:latest
$ cat /tmp/ecs.toml 
[sources.my_source]
  type = "aws_ecs_metrics"
  endpoint = "http://localhost:9080/v3/task/stats"

[sinks.my_sink]
  type = "console"
  inputs = ["my_source"]
  encoding.codec = "json"
$ ./target/debug/vector --config /tmp/ecs.toml -vvv
Nov 12 19:03:52.830  INFO vector::app: Log level is enabled. level="trace"
Nov 12 19:03:52.830  INFO vector::app: Loading configs. path=["/tmp/ecs.toml"]
Nov 12 19:03:52.845  INFO vector::topology: Running healthchecks.
Nov 12 19:03:52.845  INFO vector::topology: Starting source. name="my_source"
Nov 12 19:03:52.846  INFO vector::topology: Starting sink. name="my_sink"
Nov 12 19:03:52.846  INFO vector: Vector has started. version="0.11.0" git_version="v0.9.0-1322-g1e27ce4" released="Thu, 12 Nov 2020 18:22:59 +0000" arch="x86_64"
Nov 12 19:03:52.846  INFO vector::topology::builder: Healthcheck: Passed.
Nov 12 19:03:52.848 TRACE vector: Beep.
Nov 12 19:03:53.848 TRACE vector: Beep.
Nov 12 19:03:54.062 DEBUG source{component_kind="source" component_name=my_source component_type=aws_ecs_metrics}: vector::internal_events::aws_ecs_metrics: Request completed.
Nov 12 19:03:54.063 ERROR source{component_kind="source" component_name=my_source component_type=aws_ecs_metrics}: vector::internal_events::aws_ecs_metrics: Parsing error. url=http://localhost:9080/v3/task/stats error=missing field `name` at line 1 column 1985
Nov 12 19:03:54.063 DEBUG source{component_kind="source" component_name=my_source component_type=aws_ecs_metrics}: vector::internal_events::aws_ecs_metrics: Failed to parse response:\n\n{"01e2d287b1718a72fea24c12756dafb3c4d2760f3a872b518d29e7b908228858":{"read":"2020-11-12T19:03:54.049464618Z","preread":"2020-11-12T19:03:53.039629629Z","pids_stats":{"current":89},"blkio_stats":{"io_service_bytes_recursive":[],"io_serviced_recursive":[],"io_queue_recursive":[],"io_service_time_recursive":[],"io_wait_time_recursive":[],"io_merged_recursive":[],"io_time_recursive":[],"sectors_recursive":[]},"num_procs":0,"storage_stats":{},"cpu_stats":{"cpu_usage":{"total_usage":219281161670,"percpu_usage":[27812716238,27369612672,27640736386,26945533368,27082785942,27614184923,26876764053,27938828088],"usage_in_kernelmode":81390000000,"usage_in_usermode":106870000000},"system_cpu_usage":948006450000000,"online_cpus":8,"throttling_data":{"periods":0,"throttled_periods":0,"throttled_time":0}},"precpu_stats":{"cpu_usage":{"total_usage":219278516836,"percpu_usage":[27812660525,27369411965,27640264607,26945361987,27082672537,27613466588,26876259225,27938419402],"usage_in_kernelmode":81390000000,"usage_in_usermode":106860000000},"system_cpu_usage":947998360000000,"online_cpus":8,"throttling_data":{"periods":0,"throttled_periods":0,"throttled_time":0}},"memory_stats":{"usage":504631296,"max_usage":505192448,"stats":{"active_anon":456568832,"active_file":10936320,"cache":21663744,"dirty":4096,"hierarchical_memory_limit":9223372036854771712,"hierarchical_memsw_limit":9223372036854771712,"inactive_anon":19005440,"inactive_file":10727424,"mapped_file":13348864,"pgfault":209986,"pgmajfault":230,"pgpgin":194039,"pgpgout":72643,"rss":475574272,"rss_huge":0,"total_active_anon":456568832,"total_active_file":10936320,"total_cache":21663744,"total_dirty":4096,"total_inactive_anon":19005440,"total_inactive_file":10727424,"total_mapped_file":13348864,"total_pgfault":209986,"total_pgmajfault":230,"total_pgpgin":194039,"total_pgpgout":72643,"total_rss":475574272,"total_rss_huge":0,"total_unevictable":0,"total_writeback":0,"unevictable":0,"writeback":0},"limit":33359007744}},"28b66a2b97d5a1145e5e37dd24f76a4b288ba5cb07750d5cbf35e8a41c680b34":{"read":"2020-11-12T19:03:54.054140137Z","preread":"2020-11-12T19:03:53.042957849Z","pids_stats":{"current":63},"blkio_stats":{"io_service_bytes_recursive":[],"io_serviced_recursive":[],"io_queue_recursive":[],"io_service_time_recursive":[],"io_wait_time_recursive":[],"io_merged_recursive":[],"io_time_recursive":[],"sectors_recursive":[]},"num_procs":0,"storage_stats":{},"cpu_stats":{"cpu_usage":{"total_usage":229812006808,"percpu_usage":[31604183234,29659887938,26982983850,25729582830,32608368839,30126915237,27725430160,25374654720],"usage_in_kernelmode":71440000000,"usage_in_usermode":135930000000},"system_cpu_usage":948006480000000,"online_cpus":8,"throttling_data":{"periods":0,"throttled_periods":0,"throttled_time":0}},"precpu_stats":{"cpu_usage":{"total_usage":229810427970,"percpu_usage":[31603385973,29659698575,26982864371,25729435193,32608251154,30126879237,27725414662,25374498805],"usage_in_kernelmode":71440000000,"usage_in_usermode":135930000000},"system_cpu_usage":947998390000000,"online_cpus":8,"throttling_data":{"periods":0,"throttled_periods":0,"throttled_time":0}},"memory_stats":{"usage":165924864,"max_usage":394944512,"stats":{"active_anon":79933440,"active_file":1286144,"cache":6070272,"dirty":12288,"hierarchical_memory_limit":9223372036854771712,"hierarchical_memsw_limit":9223372036854771712,"inactive_anon":71913472,"inactive_file":4784128,"mapped_file":5382144,"pgfault":292407,"pgmajfault":800,"pgpgin":240390,"pgpgout":201836,"rss":151846912,"rss_huge":0,"total_active_anon":79933440,"total_active_file":1286144,"total_cache":6070272,"total_dirty":12288,"total_inactive_anon":71913472,"total_inactive_file":4784128,"total_mapped_file":5382144,"total_pgfault":292407,"total_pgmajfault":800,"total_pgpgin":240390,"total_pgpgout":201836,"total_rss":151846912,"total_rss_huge":0,"total_unevictable":0,"total_writeback":0,"unevictable":0,"writeback":0},"limit":33359007744}},"6caaf033fca8b4ce0af9ada18e668e851bd996bd5261d11c81229df8c13ab602":{"read":"2020-11-12T19:03:54.052554549Z","preread":"2020-11-12T19:03:53.041473098Z","pids_stats":{"current":14},"blkio_stats":{"io_service_bytes_recursive":[],"io_serviced_recursive":[],"io_queue_recursive":[],"io_service_time_recursive":[],"io_wait_time_recursive":[],"io_merged_recursive":[],"io_time_recursive":[],"sectors_recursive":[]},"num_procs":0,"storage_stats":{},"cpu_stats":{"cpu_usage":{"total_usage":7095632404,"percpu_usage":[858044913,948773155,895865404,843324783,810732310,834452396,986229362,918210081],"usage_in_kernelmode":3900000000,"usage_in_usermode":2470000000},"system_cpu_usage":948006470000000,"online_cpus":8,"throttling_data":{"periods":0,"throttled_periods":0,"throttled_time":0}},"precpu_stats":{"cpu_usage":{"total_usage":7095467418,"percpu_usage":[857939052,948773155,895865404,843324783,810732310,834413470,986209163,918210081],"usage_in_kernelmode":3900000000,"usage_in_usermode":2470000000},"system_cpu_usage":947998380000000,"online_cpus":8,"throttling_data":{"periods":0,"throttled_periods":0,"throttled_time":0}},"memory_stats":{"usage":13725696,"max_usage":16748544,"stats":{"active_anon":1454080,"active_file":6389760,"cache":8708096,"dirty":0,"hierarchical_memory_limit":9223372036854771712,"hierarchical_memsw_limit":9223372036854771712,"inactive_anon":368640,"inactive_file":4202496,"mapped_file":8019968,"pgfault":4792,"pgmajfault":161,"pgpgin":7195,"pgpgout":4164,"rss":3706880,"rss_huge":0,"total_active_anon":1454080,"total_active_file":6389760,"total_cache":8708096,"total_dirty":0,"total_inactive_anon":368640,"total_inactive_file":4202496,"total_mapped_file":8019968,"total_pgfault":4792,"total_pgmajfault":161,"total_pgpgin":7195,"total_pgpgout":4164,"total_rss":3706880,"total_rss_huge":0,"total_unevictable":0,"total_writeback":0,"unevictable":0,"writeback":0},"limit":33359007744}},"7145c82e25394eea5bfda76b6a0e628f46ed9193c1c2d0ac3531fb8aa49e428b":{"read":"2020-11-12T19:03:54.05973861Z","preread":"2020-11-12T19:03:53.035613072Z","pids_stats":{"current":27},"blkio_stats":{"io_service_bytes_recursive":[],"io_serviced_recursive":[],"io_queue_recursive":[],"io_service_time_recursive":[],"io_wait_time_recursive":[],"io_merged_recursive":[],"io_time_recursive":[],"sectors_recursive":[]},"num_procs":0,"storage_stats":{},"cpu_stats":{"cpu_usage":{"total_usage":38521720466,"percpu_usage":[4946498380,4891746462,4650303028,4481880089,5054996402,4674030733,5001958092,4820307280],"usage_in_kernelmode":15210000000,"usage_in_usermode":20850000000},"system_cpu_usage":948006530000000,"online_cpus":8,"throttling_data":{"periods":0,"throttled_periods":0,"throttled_time":0}},"precpu_stats":{"cpu_usage":{"total_usage":38521247099,"percpu_usage":[4946498380,4891746462,4649885335,4481845699,5054975118,4674030733,5001958092,4820307280],"usage_in_kernelmode":15210000000,"usage_in_usermode":20850000000},"system_cpu_usage":947998350000000,"online_cpus":8,"throttling_data":{"periods":0,"throttled_periods":0,"throttled_time":0}},"memory_stats":{"usage":89194496,"max_usage":101527552,"stats":{"active_anon":50008064,"active_file":9068544,"cache":18239488,"dirty":8192,"hierarchical_memory_limit":9223372036854771712,"hierarchical_memsw_limit":9223372036854771712,"inactive_anon":16388096,"inactive_file":9170944,"mapped_file":16289792,"pgfault":30870,"pgmajfault":207,"pgpgin":33143,"pgpgout":12480,"rss":66396160,"rss_huge":0,"total_active_anon":50008064,"total_active_file":9068544,"total_cache":18239488,"total_dirty":8192,"total_inactive_anon":16388096,"total_inactive_file":9170944,"total_mapped_file":16289792,"total_pgfault":30870,"total_pgmajfault":207,"total_pgpgin":33143,"total_pgpgout":12480,"total_rss":66396160,"total_rss_huge":0,"total_unevictable":0,"total_writeback":0,"unevictable":0,"writeback":0},"limit":33359007744}},"7e363eaa295edfce8135712ea31832a82b16da78b84c5780ea74ee8ec1c54c07":{"read":"2020-11-12T19:03:54.055630374Z","preread":"2020-11-12T19:03:53.044461193Z","pids_stats":{"current":2},"blkio_stats":{"io_service_bytes_recursive":[],"io_serviced_recursive":[],"io_queue_recursive":[],"io_service_time_recursive":[],"io_wait_time_recursive":[],"io_merged_recursive":[],"io_time_recursive":[],"sectors_recursive":[]},"num_procs":0,"storage_stats":{},"cpu_stats":{"cpu_usage":{"total_usage":248219219,"percpu_usage":[3817921,227624078,6226159,3631274,6213949,275506,404213,26119],"usage_in_kernelmode":10000000,"usage_in_usermode":210000000},"system_cpu_usage":948006490000000,"online_cpus":8,"throttling_data":{"periods":0,"throttled_periods":0,"throttled_time":0}},"precpu_stats":{"cpu_usage":{"total_usage":248219219,"percpu_usage":[3817921,227624078,6226159,3631274,6213949,275506,404213,26119],"usage_in_kernelmode":10000000,"usage_in_usermode":210000000},"system_cpu_usage":947998400000000,"online_cpus":8,"throttling_data":{"periods":0,"throttled_periods":0,"throttled_time":0}},"memory_stats":{"usage":1380352,"max_usage":6656000,"stats":{"active_anon":28672,"active_file":24576,"cache":151552,"dirty":0,"hierarchical_memory_limit":9223372036854771712,"hierarchical_memsw_limit":9223372036854771712,"inactive_anon":147456,"inactive_file":126976,"mapped_file":151552,"pgfault":1926,"pgmajfault":5,"pgpgin":1711,"pgpgout":1631,"rss":176128,"rss_huge":0,"total_active_anon":28672,"total_active_file":24576,"total_cache":151552,"total_dirty":0,"total_inactive_anon":147456,"total_inactive_file":126976,"total_mapped_file":151552,"total_pgfault":1926,"total_pgmajfault":5,"total_pgpgin":1711,"total_pgpgout":1631,"total_rss":176128,"total_rss_huge":0,"total_unevictable":0,"total_writeback":0,"unevictable":0,"writeback":0},"limit":33359007744}},"ca636c261142a191f60b497fec1e17215770dd13829066e6647ffb6f1a07fccf":{"read":"2020-11-12T19:03:54.047879526Z","preread":"2020-11-12T19:03:53.037393336Z","pids_stats":{"current":8},"blkio_stats":{"io_service_bytes_recursive":[],"io_serviced_recursive":[],"io_queue_recursive":[],"io_service_time_recursive":[],"io_wait_time_recursive":[],"io_merged_recursive":[],"io_time_recursive":[],"sectors_recursive":[]},"num_procs":0,"storage_stats":{},"cpu_stats":{"cpu_usage":{"total_usage":290685420,"percpu_usage":[9663244,9025619,228762130,12922305,10459705,1284562,6970662,11597193],"usage_in_kernelmode":40000000,"usage_in_usermode":230000000},"system_cpu_usage":948006440000000,"online_cpus":8,"throttling_data":{"periods":0,"throttled_periods":0,"throttled_time":0}},"precpu_stats":{"cpu_usage":{"total_usage":290685420,"percpu_usage":[9663244,9025619,228762130,12922305,10459705,1284562,6970662,11597193],"usage_in_kernelmode":40000000,"usage_in_usermode":230000000},"system_cpu_usage":947998360000000,"online_cpus":8,"throttling_data":{"periods":0,"throttled_periods":0,"throttled_time":0}},"memory_stats":{"usage":2248704,"max_usage":6606848,"stats":{"active_anon":524288,"active_file":20480,"cache":61440,"dirty":0,"hierarchical_memory_limit":9223372036854771712,"hierarchical_memsw_limit":9223372036854771712,"inactive_anon":614400,"inactive_file":40960,"mapped_file":61440,"pgfault":2572,"pgmajfault":6,"pgpgin":2192,"pgpgout":1899,"rss":1138688,"rss_huge":0,"total_active_anon":524288,"total_active_file":20480,"total_cache":61440,"total_dirty":0,"total_inactive_anon":614400,"total_inactive_file":40960,"total_mapped_file":61440,"total_pgfault":2572,"total_pgmajfault":6,"total_pgpgin":2192,"total_pgpgout":1899,"total_rss":1138688,"total_rss_huge":0,"total_unevictable":0,"total_writeback":0,"unevictable":0,"writeback":0},"limit":33359007744}},"cc5ee81ef971a255485bab168fa890a66d02377d7264b667f22091d9dcf76395":{"read":"2020-11-12T19:03:54.057865037Z","preread":"2020-11-12T19:03:53.045860625Z","pids_stats":{"current":12},"blkio_stats":{"io_service_bytes_recursive":[],"io_serviced_recursive":[],"io_queue_recursive":[],"io_service_time_recursive":[],"io_wait_time_recursive":[],"io_merged_recursive":[],"io_time_recursive":[],"sectors_recursive":[]},"num_procs":0,"storage_stats":{},"cpu_stats":{"cpu_usage":{"total_usage":300213110,"percpu_usage":[22431628,13311216,21831879,13575186,22263383,16745653,174908105,15146060],"usage_in_kernelmode":10000000,"usage_in_usermode":200000000},"system_cpu_usage":948006530000000,"online_cpus":8,"throttling_data":{"periods":0,"throttled_periods":0,"throttled_time":0}},"precpu_stats":{"cpu_usage":{"total_usage":298452944,"percpu_usage":[21834920,13079192,21257005,13521761,22263383,16524722,174853129,15118832],"usage_in_kernelmode":10000000,"usage_in_usermode":200000000},"system_cpu_usage":947998440000000,"online_cpus":8,"throttling_data":{"periods":0,"throttled_periods":0,"throttled_time":0}},"memory_stats":{"usage":7507968,"max_usage":7757824,"stats":{"active_anon":3678208,"active_file":0,"cache":0,"dirty":0,"hierarchical_memory_limit":9223372036854771712,"hierarchical_memsw_limit":9223372036854771712,"inactive_anon":0,"inactive_file":2449408,"mapped_file":0,"pgfault":4037,"pgmajfault":0,"pgpgin":3351,"pgpgout":1847,"rss":6160384,"rss_huge":0,"total_active_anon":3678208,"total_active_file":0,"total_cache":0,"total_dirty":0,"total_inactive_anon":0,"total_inactive_file":2449408,"total_mapped_file":0,"total_pgfault":4037,"total_pgmajfault":0,"total_pgpgin":3351,"total_pgpgout":1847,"total_rss":6160384,"total_rss_huge":0,"total_unevictable":0,"total_writeback":0,"unevictable":0,"writeback":0},"limit":33359007744}}}
\n\n url=http://localhost:9080/v3/task/stats rate_limit_secs=10
```

This brings me to another request, could we add an integration test for this using that docker container? You can see this NATS sink PR for an example: https://github.com/timberio/vector/pull/3605 . I would just verify that it can request and parse the v2, v3, and v4 endpoints.

**Comment by drunkirishcoder** [comment]
OK I added a test to the AWS integration test suite. Because the local ECS container is scraping live metrics from the host and it takes a bit time, I had to up the delay to 5s to reliably get it passing on my machine. Not sure what else to do there. All comments are addressed.

**Comment by drunkirishcoder** [comment]
Had a discord discussion with @jszwedko about future enhancement to metrics metadata. Decided to change the auto endpoint detection slightly. I think this PR is complete now.

**Comment by jamtur01** [comment]
@jszwedko You a 💚 on this? If so, merge away.

**Review by jszwedko — APPROVED** [review]
Great work on this @drunkirishcoder !

I pushed one more commit adding integration tests for v2 and v4.

**Comment by drunkirishcoder** [comment]
great thanks for the help!

**Pull request merged**